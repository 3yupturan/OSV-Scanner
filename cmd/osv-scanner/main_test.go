// main cannot be accessed directly, so cannot use main_test
package main

import (
	"bytes"
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
	"testing"
)

func dedent(t *testing.T, str string) string {
	t.Helper()

	// 0. replace all tabs with spaces
	str = strings.ReplaceAll(str, "\t", "  ")

	// 1. remove trailing whitespace
	re := regexp.MustCompile(`\r?\n([\t ]*)$`)
	str = re.ReplaceAllString(str, "")

	// 2. if any of the lines are not indented, return as we're already dedent-ed
	re = regexp.MustCompile(`(^|\r?\n)[^\t \n]`)
	if re.MatchString(str) {
		return str
	}

	// 3. find all line breaks to determine the highest common indentation level
	re = regexp.MustCompile(`\n[\t ]+`)
	matches := re.FindAllString(str, -1)

	// 4. remove the common indentation from all strings
	if matches != nil {
		size := len(matches[0]) - 1

		for _, match := range matches {
			if len(match)-1 < size {
				size = len(match) - 1
			}
		}

		re := regexp.MustCompile(`\n[\t ]{` + fmt.Sprint(size) + `}`)
		str = re.ReplaceAllString(str, "\n")
	}

	// 5. Remove leading whitespace.
	re = regexp.MustCompile(`^\r?\n`)
	str = re.ReplaceAllString(str, "")

	return str
}

// checks if two strings are equal, treating any occurrences of `%%` in the
// expected string to mean "any text"
func areEqual(t *testing.T, actual, expect string) bool {
	t.Helper()

	expect = regexp.QuoteMeta(expect)
	expect = strings.ReplaceAll(expect, "%%", ".+")

	re := regexp.MustCompile(`^` + expect + `$`)

	return re.MatchString(actual)
}

type cliTestCase struct {
	name         string
	args         []string
	wantExitCode int
	wantStdout   string
	wantStderr   string
}

func testCli(t *testing.T, tc cliTestCase) {
	t.Helper()

	stdoutBuffer := &bytes.Buffer{}
	stderrBuffer := &bytes.Buffer{}

	ec := run(tc.args, stdoutBuffer, stderrBuffer)
	// ec := run(tc.args, os.Stdout, os.Stderr)

	stdout := stdoutBuffer.String()
	stderr := stderrBuffer.String()

	if ec != tc.wantExitCode {
		t.Errorf("cli exited with code %d, not %d", ec, tc.wantExitCode)
	}

	if !areEqual(t, dedent(t, stdout), dedent(t, tc.wantStdout)) {
		t.Errorf("stdout\n got:\n%s\n\n want:\n%s", dedent(t, stdout), dedent(t, tc.wantStdout))
	}

	if !areEqual(t, dedent(t, stderr), dedent(t, tc.wantStderr)) {
		t.Errorf("stderr\n got:\n%s\n\n want:\n%s", dedent(t, stderr), dedent(t, tc.wantStderr))
	}
}

func TestRun(t *testing.T) {
	t.Parallel()

	tests := []cliTestCase{
		{
			name:         "",
			args:         []string{""},
			wantExitCode: 128,
			wantStdout:   "",
			wantStderr: `
        No package sources found, --help for usage information.
			`,
		},
		{
			name:         "",
			args:         []string{"", "--version"},
			wantExitCode: 0,
			wantStdout: `
				osv-scanner version: dev
				commit: n/a
				built at: n/a
			`,
			wantStderr: "",
		},
		// one specific supported lockfile
		{
			name:         "",
			args:         []string{"", "./fixtures/locks-many/composer.lock"},
			wantExitCode: 0,
			wantStdout: `
				Scanning dir ./fixtures/locks-many/composer.lock
        Scanned %%/fixtures/locks-many/composer.lock file and found 1 packages
			`,
			wantStderr: "",
		},
		// one specific unsupported lockfile
		{
			name:         "",
			args:         []string{"", "./fixtures/locks-many/not-a-lockfile.toml"},
			wantExitCode: 128,
			wantStdout: `
				Scanning dir ./fixtures/locks-many/not-a-lockfile.toml
			`,
			wantStderr: `
				No package sources found, --help for usage information.
			`,
		},
		// all supported lockfiles in the directory should be checked
		{
			name:         "",
			args:         []string{"", "./fixtures/locks-many"},
			wantExitCode: 0,
			wantStdout: `
				Scanning dir ./fixtures/locks-many
				Scanned %%/fixtures/locks-many/Gemfile.lock file and found 1 packages
				Scanned %%/fixtures/locks-many/composer.lock file and found 1 packages
				Scanned %%/fixtures/locks-many/yarn.lock file and found 1 packages
			`,
			wantStderr: "",
		},
		// all supported lockfiles in the directory should be checked
		{
			name:         "",
			args:         []string{"", "./fixtures/locks-many-with-invalid"},
			wantExitCode: 127,
			wantStdout: `
				Scanning dir ./fixtures/locks-many-with-invalid
				Scanned %%/fixtures/locks-many-with-invalid/Gemfile.lock file and found 1 packages
				Scanned %%/fixtures/locks-many-with-invalid/yarn.lock file and found 1 packages
			`,
			wantStderr: `
				Attempted to scan lockfile but failed: %%/fixtures/locks-many-with-invalid/composer.lock
			`,
		},
		// only the files in the given directories are checked by default (no recursion)
		{
			name:         "",
			args:         []string{"", "./fixtures/locks-one-with-nested"},
			wantExitCode: 0,
			wantStdout: `
				Scanning dir ./fixtures/locks-one-with-nested
				Scanned %%/fixtures/locks-one-with-nested/yarn.lock file and found 1 packages
			`,
			wantStderr: "",
		},
		// nested directories are checked when `--recursive` is passed
		{
			name:         "",
			args:         []string{"", "--recursive", "./fixtures/locks-one-with-nested"},
			wantExitCode: 0,
			wantStdout: `
				Scanning dir ./fixtures/locks-one-with-nested
				Scanned %%/fixtures/locks-one-with-nested/nested/composer.lock file and found 1 packages
				Scanned %%/fixtures/locks-one-with-nested/yarn.lock file and found 1 packages
			`,
			wantStderr: "",
		},
		// output with json
		{
			name:         "",
			args:         []string{"", "--json", "./fixtures/locks-many/composer.lock"},
			wantExitCode: 0,
			wantStdout: `
				{
					"results": []
				}
			`,
			wantStderr: `
				Scanning dir ./fixtures/locks-many/composer.lock
				Scanned %%/fixtures/locks-many/composer.lock file and found 1 packages
			`,
		},
		{
			name:         "",
			args:         []string{"", "--format", "json", "./fixtures/locks-many/composer.lock"},
			wantExitCode: 0,
			wantStdout: `
				{
					"results": []
				}
			`,
			wantStderr: `
				Scanning dir ./fixtures/locks-many/composer.lock
				Scanned %%/fixtures/locks-many/composer.lock file and found 1 packages
			`,
		},
		// output format: markdown table
		{
			name:         "",
			args:         []string{"", "--format", "markdown", "./fixtures/locks-many/composer.lock"},
			wantExitCode: 0,
			wantStdout: `
				Scanning dir ./fixtures/locks-many/composer.lock
				Scanned %%/fixtures/locks-many/composer.lock file and found 1 packages
			`,
			wantStderr: "",
		},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			testCli(t, tt)
		})
	}
}

func TestRun_ParseAs(t *testing.T) {
	t.Parallel()

	tests := []cliTestCase{
		// invalid parse-as
		{
			name:         "",
			args:         []string{"", "--parse-as", "my-file"},
			wantExitCode: 127,
			wantStdout:   "",
			wantStderr: `
				Don't know how to parse files as "my-file" - supported values are:
					buildscript-gradle.lockfile
					Cargo.lock
					composer.lock
					conan.lock
					Gemfile.lock
					go.mod
					gradle.lockfile
					mix.lock
					package-lock.json
					packages.lock.json
					Pipfile.lock
					pnpm-lock.yaml
					poetry.lock
					pom.xml
					pubspec.lock
					requirements.txt
					yarn.lock

			`,
		},
		// when a path to a file is given, parse-as is applied to that file
		{
			name:         "",
			args:         []string{"", "--parse-as", "package-lock.json", filepath.FromSlash("./fixtures/locks-insecure/my-package-lock.json")},
			wantExitCode: 1,
			wantStdout: `
				Scanning dir ./fixtures/locks-insecure/my-package-lock.json
				Scanned %%/fixtures/locks-insecure/my-package-lock.json file and found 1 packages
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
				| OSV URL (ID IN BOLD)                              | ECOSYSTEM | PACKAGE   | VERSION | SOURCE                                       |
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
				| https://osv.dev/vulnerability/GHSA-whgm-jr23-g3j9 | npm       | ansi-html | 0.0.1   | fixtures/locks-insecure/my-package-lock.json |
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
			`,
			wantStderr: "",
		},
		// when a path to a directory is given, parse-as is applied to all files in the directory
		{
			name:         "",
			args:         []string{"", "--parse-as", "package-lock.json", filepath.FromSlash("./fixtures/locks-insecure")},
			wantExitCode: 1,
			wantStdout: `
				Scanning dir ./fixtures/locks-insecure
				Scanned %%/fixtures/locks-insecure/composer.lock file and found 0 packages
				Scanned %%/fixtures/locks-insecure/my-package-lock.json file and found 1 packages
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
				| OSV URL (ID IN BOLD)                              | ECOSYSTEM | PACKAGE   | VERSION | SOURCE                                       |
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
				| https://osv.dev/vulnerability/GHSA-whgm-jr23-g3j9 | npm       | ansi-html | 0.0.1   | fixtures/locks-insecure/my-package-lock.json |
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
			`,
			wantStderr: "",
		},
		// files that error on parsing don't stop parsable files from being checked
		{
			name:         "",
			args:         []string{"", "--parse-as", "package-lock.json", filepath.FromSlash("./fixtures/locks-empty")},
			wantExitCode: 128,
			wantStdout: `
				Scanning dir ./fixtures/locks-empty
				Scanned %%/fixtures/locks-empty/composer.lock file and found 0 packages
			`,
			wantStderr: `
				Attempted to scan lockfile but failed: %%/fixtures/locks-empty/Gemfile.lock
				Attempted to scan lockfile but failed: %%/fixtures/locks-empty/yarn.lock
				No package sources found, --help for usage information.
			`,
		},
		// files that error on parsing don't stop parsable files from being checked
		{
			name:         "",
			args:         []string{"", "--parse-as", "package-lock.json", filepath.FromSlash("./fixtures/locks-empty"), filepath.FromSlash("./fixtures/locks-insecure")},
			wantExitCode: 1,
			wantStdout: `
				Scanning dir ./fixtures/locks-empty
				Scanned %%/fixtures/locks-empty/composer.lock file and found 0 packages
				Scanning dir ./fixtures/locks-insecure
				Scanned %%/fixtures/locks-insecure/composer.lock file and found 0 packages
				Scanned %%/fixtures/locks-insecure/my-package-lock.json file and found 1 packages
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
				| OSV URL (ID IN BOLD)                              | ECOSYSTEM | PACKAGE   | VERSION | SOURCE                                       |
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
				| https://osv.dev/vulnerability/GHSA-whgm-jr23-g3j9 | npm       | ansi-html | 0.0.1   | fixtures/locks-insecure/my-package-lock.json |
				+---------------------------------------------------+-----------+-----------+---------+----------------------------------------------+
			`,
			wantStderr: `
				Attempted to scan lockfile but failed: %%/fixtures/locks-empty/Gemfile.lock
				Attempted to scan lockfile but failed: %%/fixtures/locks-empty/yarn.lock
			`,
		},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			testCli(t, tt)
		})
	}
}
