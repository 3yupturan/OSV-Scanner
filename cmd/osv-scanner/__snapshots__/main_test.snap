
[TestRun2/#00 - 1]

---

[TestRun2/#00 - 2]
No package sources found, --help for usage information.

---

[TestRun2/#05 - 1]

---

[TestRun2/#05 - 2]
unsupported output format "unknown" - must be one of: table, json, markdown, sarif, gh-annotations

---

[TestRun2/#01 - 1]
Scanning dir ./fixtures/locks-many/not-a-lockfile.toml

---

[TestRun2/#01 - 2]
No package sources found, --help for usage information.

---

[TestRun2/Empty_gh-annotations_output - 1]

---

[TestRun2/Empty_gh-annotations_output - 2]
Scanning dir ./fixtures/locks-many/composer.lock
Scanned <rootdir>/fixtures/locks-many/composer.lock file and found 1 package

---

[TestRun2/json_output_2 - 1]
{
  "results": [],
  "experimental_config": {
    "licenses": {
      "summary": false,
      "allowlist": null
    }
  }
}

---

[TestRun2/json_output_2 - 2]
Scanning dir ./fixtures/locks-many/composer.lock
Scanned <rootdir>/fixtures/locks-many/composer.lock file and found 1 package

---

[TestRun2/json_output_1 - 1]
{
  "results": [],
  "experimental_config": {
    "licenses": {
      "summary": false,
      "allowlist": null
    }
  }
}

---

[TestRun2/json_output_1 - 2]
Scanning dir ./fixtures/locks-many/composer.lock
Scanned <rootdir>/fixtures/locks-many/composer.lock file and found 1 package

---

[TestRun2/#03 - 1]
Scanning dir ./fixtures/locks-gitignore
Scanned <rootdir>/fixtures/locks-gitignore/Gemfile.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/composer.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/ignored/Gemfile.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/ignored/yarn.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/subdir/Gemfile.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/subdir/composer.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/subdir/yarn.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/yarn.lock file and found 1 package
No issues found

---

[TestRun2/#03 - 2]

---

[TestRun2/#02 - 1]
Scanning dir ./fixtures/locks-gitignore
Scanned <rootdir>/fixtures/locks-gitignore/Gemfile.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-gitignore/subdir/yarn.lock file and found 1 package
No issues found

---

[TestRun2/#02 - 2]

---

[TestRun2/only_the_files_in_the_given_directories_are_checked_by_default_(no_recursion) - 1]
Scanning dir ./fixtures/locks-one-with-nested
Scanned <rootdir>/fixtures/locks-one-with-nested/yarn.lock file and found 1 package
No issues found

---

[TestRun2/only_the_files_in_the_given_directories_are_checked_by_default_(no_recursion) - 2]

---

[TestRun2/nested_directories_are_checked_when_`--recursive`_is_passed - 1]
Scanning dir ./fixtures/locks-one-with-nested
Scanned <rootdir>/fixtures/locks-one-with-nested/nested/composer.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-one-with-nested/yarn.lock file and found 1 package
No issues found

---

[TestRun2/nested_directories_are_checked_when_`--recursive`_is_passed - 2]

---

[TestRun2/#04 - 1]
Scanning dir ./fixtures/locks-many/package-lock.json
Scanned <rootdir>/fixtures/locks-many/package-lock.json file and found 1 package
| OSV URL | CVSS | Ecosystem | Package | Version | Source |
| --- | --- | --- | --- | --- | --- |
| https://osv.dev/GHSA-whgm-jr23-g3j9 | 7.5 | npm | ansi-html | 0.0.1 | fixtures/locks-many/package-lock.json |

---

[TestRun2/#04 - 2]

---

[TestRun2/all_supported_lockfiles_in_the_directory_should_be_checked - 1]
Scanning dir ./fixtures/locks-many-with-invalid
Scanned <rootdir>/fixtures/locks-many-with-invalid/Gemfile.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-many-with-invalid/yarn.lock file and found 1 package

---

[TestRun2/gh-annotations_with_vulns - 1]

---

[TestRun2/all_supported_lockfiles_in_the_directory_should_be_checked - 2]
Attempted to scan lockfile but failed: <rootdir>/fixtures/locks-many-with-invalid/composer.lock

---

[TestRun2/gh-annotations_with_vulns - 2]
Scanning dir ./fixtures/locks-many/package-lock.json
Scanned <rootdir>/fixtures/locks-many/package-lock.json file and found 1 package
::error file=fixtures/locks-many/package-lock.json::fixtures/locks-many/package-lock.json%0A+-----------+-------------------------------------+------+-----------------+---------------+%0A| PACKAGE   | VULNERABILITY ID                    | CVSS | CURRENT VERSION | FIXED VERSION |%0A+-----------+-------------------------------------+------+-----------------+---------------+%0A| ansi-html | https://osv.dev/GHSA-whgm-jr23-g3j9 | 7.5  | 0.0.1           | 0.0.8         |%0A+-----------+-------------------------------------+------+-----------------+---------------+
---

[TestRun2/one_specific_supported_lockfile - 1]
Scanning dir ./fixtures/locks-many/composer.lock
Scanned <rootdir>/fixtures/locks-many/composer.lock file and found 1 package
No issues found

---

[TestRun2/one_specific_supported_lockfile - 2]

---

[TestRun2/one_specific_supported_lockfile_with_ignore - 1]
Scanning dir ./fixtures/locks-test-ignore/package-lock.json
Scanned <rootdir>/fixtures/locks-test-ignore/package-lock.json file and found 1 package
Loaded filter from: <rootdir>/fixtures/locks-test-ignore/osv-scanner.toml
CVE-2021-23424 and 1 alias have been filtered out because: Test manifest file (alpine.cdx.xml)
Filtered 1 vulnerability from output
No issues found

---

[TestRun2/one_specific_supported_lockfile_with_ignore - 2]

---

[TestRun2/one_specific_supported_sbom_with_vulns - 1]
Scanned <rootdir>/fixtures/sbom-insecure/alpine.cdx.xml as CycloneDX SBOM and found 15 packages
+--------------------------------+------+-----------+---------+------------+---------------------------------------+
| OSV URL                        | CVSS | ECOSYSTEM | PACKAGE | VERSION    | SOURCE                                |
+--------------------------------+------+-----------+---------+------------+---------------------------------------+
| https://osv.dev/CVE-2022-48174 | 9.8  | Alpine    | busybox | 1.35.0-r29 | fixtures/sbom-insecure/alpine.cdx.xml |
| https://osv.dev/CVE-2022-37434 | 9.8  | Alpine    | zlib    | 1.2.10-r2  | fixtures/sbom-insecure/alpine.cdx.xml |
+--------------------------------+------+-----------+---------+------------+---------------------------------------+

---

[TestRun2/one_specific_supported_sbom_with_vulns - 2]

---

[TestRun2/Scan_locks-many - 1]
Scanning dir ./fixtures/locks-many
Scanned <rootdir>/fixtures/locks-many/Gemfile.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-many/alpine.cdx.xml as CycloneDX SBOM and found 15 packages
Scanned <rootdir>/fixtures/locks-many/composer.lock file and found 1 package
Scanned <rootdir>/fixtures/locks-many/package-lock.json file and found 1 package
Scanned <rootdir>/fixtures/locks-many/yarn.lock file and found 1 package
Loaded filter from: <rootdir>/fixtures/locks-many/osv-scanner.toml
CVE-2022-48174 has been filtered out because: Test manifest file (alpine.cdx.xml)
GHSA-whgm-jr23-g3j9 and 1 alias have been filtered out because: Test manifest file
Filtered 2 vulnerabilities from output
No issues found

---

[TestRun2/Scan_locks-many - 2]

---

[TestRun2/folder_of_supported_sbom_with_vulns - 1]
Scanning dir ./fixtures/sbom-insecure/
Scanned <rootdir>/fixtures/sbom-insecure/alpine.cdx.xml as CycloneDX SBOM and found 15 packages
Scanned <rootdir>/fixtures/sbom-insecure/postgres-stretch.cdx.xml as CycloneDX SBOM and found 136 packages
+-------------------------------------+------+-----------+--------------------------------+------------------------------------+-------------------------------------------------+
| OSV URL                             | CVSS | ECOSYSTEM | PACKAGE                        | VERSION                            | SOURCE                                          |
+-------------------------------------+------+-----------+--------------------------------+------------------------------------+-------------------------------------------------+
| https://osv.dev/CVE-2022-48174      | 9.8  | Alpine    | busybox                        | 1.35.0-r29                         | fixtures/sbom-insecure/alpine.cdx.xml           |
| https://osv.dev/CVE-2022-37434      | 9.8  | Alpine    | zlib                           | 1.2.10-r2                          | fixtures/sbom-insecure/alpine.cdx.xml           |
| https://osv.dev/DLA-3022-1          |      | Debian    | dpkg                           | 1.18.25                            | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/GHSA-v95c-p5hm-xq8f | 6    | Go        | github.com/opencontainers/runc | v1.0.1                             | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/GO-2022-0274        |      |           |                                |                                    |                                                 |
| https://osv.dev/GHSA-f3fp-gc8g-vw66 | 5.9  | Go        | github.com/opencontainers/runc | v1.0.1                             | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/GHSA-g2j6-57v7-gm8c | 6.1  | Go        | github.com/opencontainers/runc | v1.0.1                             | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/GHSA-m8cg-xc2p-r3fc | 2.5  | Go        | github.com/opencontainers/runc | v1.0.1                             | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/GHSA-vpvm-3wq2-2wvm | 7    | Go        | github.com/opencontainers/runc | v1.0.1                             | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/GHSA-p782-xgp4-8hr8 | 5.3  | Go        | golang.org/x/sys               | v0.0.0-20210817142637-7d9622a276b7 | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/GO-2022-0493        |      |           |                                |                                    |                                                 |
| https://osv.dev/DLA-3012-1          |      | Debian    | libxml2                        | 2.9.4+dfsg1-2.2+deb9u6             | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/DLA-3008-1          |      | Debian    | openssl                        | 1.1.0l-1~deb9u5                    | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
| https://osv.dev/DLA-3051-1          |      | Debian    | tzdata                         | 2021a-0+deb9u3                     | fixtures/sbom-insecure/postgres-stretch.cdx.xml |
+-------------------------------------+------+-----------+--------------------------------+------------------------------------+-------------------------------------------------+

---

[TestRun2/folder_of_supported_sbom_with_vulns - 2]

---
